File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def solution(enroll, referral, seller, amount):
    parent = dict(zip(enroll, referral))
    total = {name: 0 for name in enroll}
    
    for i in range(len(seller)):
        sales = amount[i] * 100
        cur = seller[i]
        while sales > 0 and cur != "-":
            total[cur] += sales - (sales//10)
            cur = parent[cur]
            sales //= 10
            
    return [total[name] for name in enroll]

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
    const parent = {};
    const total = {};

    for (let i = 0; i < enroll.length; i++) {
        parent[enroll[i]] = referral[i];
        total[enroll[i]] = 0;
    }

    for (let i = 0; i < seller.length; i++) {
        let sales = amount[i] * 100;
        let cur = seller[i];

        while (sales > 0 && cur !== "-") {
            total[cur] += sales - Math.floor(sales / 10);
            cur = parent[cur];
            sales = Math.floor(sales / 10);
        }
    }

    return enroll.map(name => total[name]);
}
